### PR TITLE
#18939 adding an apps-system folder

### PR DESCRIPTION
**Pull request by fabrizzio-dotCMS**
This is an update to allows separating apps in system apps (apps packaged with dotCMS) and user apps. 
Any app uploaded by the user.

### CODE REVIEW

File: dotCMS/src/integration-test/java/com/dotcms/security/apps/AppsAPIImplTest.java [hunk 1]
```diff
@@ -865,4 +866,103 @@ private boolean isSecretDestroyed(final char [] chars){
...
+        appsCache.invalidateDescriptorsCache();
+        final List<AppDescriptor> appDescriptors = api.getAppDescriptors(admin);
+
+        //Verify the file we just submitted is recognized as a system-app-file
+        final Optional<AppDescriptor> optional = appDescriptors.stream()
+                .filter(appDescriptor -> dataGen.getKey().equals(appDescriptor.getKey())).findFirst();
+        Assert.assertTrue(optional.isPresent());
+        final AppDescriptor descriptor = optional.get();
+        final AppDescriptorImpl impl = (AppDescriptorImpl)descriptor;
+        Assert.assertTrue(impl.isSystemApp());
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/security/apps/AppsAPIImplTest.java:910]: Codacy found an issue: [Unnecessary use of fully qualified name 'Assert.assertTrue' due to existing static import 'org.junit.Assert.assertTrue'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5903396)

File: dotCMS/src/main/java/com/dotcms/security/apps/AppsAPIImpl.java [hunk 2]
```diff
@@ -681,28 +688,60 @@ private void invalidateCache() {
...
-    private Set<String> listAvailableYamlFiles() throws IOException {
-        final String ymlFilesPath = getServiceDescriptorDirectory();
-        final File basePath = new File(ymlFilesPath);
+    /**
+     * This is the Apps-System-Folder which is meant to hold system apps.
+     * Those that can not be override and are always available.
+     * @return
+     */
+    static Path getSystemAppsDescriptorDirectory() throws URISyntaxException, IOException {
+        final URL res = AppsAPIImpl.class.getClassLoader().getResource("apps");
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/security/apps/AppsAPIImpl.java:709]: Codacy found an issue: [In J2EE, getClassLoader() might not work as expected.  Use Thread.currentThread().getContextClassLoader() instead.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5903396)

File: dotCMS/src/integration-test/java/com/dotcms/security/apps/AppsAPIImplTest.java [hunk 3]
```diff
@@ -865,4 +866,103 @@ private boolean isSecretDestroyed(final char [] chars){
...
+        final AppsCache appsCache = CacheLocator.getAppsCache();
+
+        //Invalidate cache so the new descriptors get picked
+        appsCache.invalidateDescriptorsCache();
+        final List<AppDescriptor> appDescriptors = api.getAppDescriptors(admin);
+
+        //Verify the file we just submitted is recognized as a system-app-file
+        final Optional<AppDescriptor> optional = appDescriptors.stream()
+                .filter(appDescriptor -> dataGen.getKey().equals(appDescriptor.getKey())).findFirst();
+        Assert.assertTrue(optional.isPresent());
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/security/apps/AppsAPIImplTest.java:907]: Codacy found an issue: [Unnecessary use of fully qualified name 'Assert.assertTrue' due to existing static import 'org.junit.Assert.assertTrue'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5903396)

File: dotCMS/src/main/java/com/dotcms/security/apps/AppsAPIImpl.java [hunk 4]
```diff
@@ -557,13 +559,18 @@ public void removeApp(final String key, final User user,
      * @param descriptor
      * @throws DotDataException
      */
-    private void removeDescriptor(final AppDescriptor descriptor) throws DotDataException{
-        final String fileName = ((AppDescriptorImpl)descriptor).getFileName();
+    private void removeDescriptor(final AppDescriptor descriptor)
+            throws DotDataException, DotSecurityException {
+        final AppDescriptorImpl appDescriptor = ((AppDescriptorImpl)descriptor);
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/security/apps/AppsAPIImpl.java:564]: Codacy found an issue: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5903396)

File: dotCMS/src/integration-test/java/com/dotcms/security/apps/AppsAPIImplTest.java [hunk 5]
```diff
@@ -865,4 +866,103 @@ private boolean isSecretDestroyed(final char [] chars){
...
+                .stringParam("p3", false,  true)
+                .withName("system-app-example")
+                .withDescription("system-app-demo")
+                .withExtraParameters(false);
+        final File file = dataGen.nextPersistedDescriptor();
+
+        //Move the file to the system folder
+        final Path systemAppsDescriptorDirectory = AppsAPIImpl.getSystemAppsDescriptorDirectory();
+        final boolean result = file.renameTo(new File(systemAppsDescriptorDirectory.toString() + File.separator + file.getName()));
+        Assert.assertTrue(result);
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/security/apps/AppsAPIImplTest.java:894]: Codacy found an issue: [Unnecessary use of fully qualified name 'Assert.assertTrue' due to existing static import 'org.junit.Assert.assertTrue'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5903396)

File: dotCMS/src/main/java/com/dotcms/security/apps/AppDescriptorImpl.java [hunk 6]
```diff
@@ -28,13 +31,15 @@
      * @param allowExtraParameters
      */
     public AppDescriptorImpl(final String fileName,
+            final boolean systemApp,
             final String name,
             final String description,
             final String iconUrl,
             final Boolean allowExtraParameters,
             final Map<String, ParamDescriptor> params) {
         super(name, description, iconUrl, allowExtraParameters, params);
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotcms/security/apps/AppDescriptorImpl.java:40]: Lots of parameters in the constructors - looks like a good candidate for a Builder pattern

Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotcms/security/apps/AppDescriptorImpl.java:43]: This constructor is public for the sake of being used on the DataGen (which is really a builder)
The real constructor is this:
` AppDescriptorImpl(final String fileName, final boolean systemApp, final AppSchema appSchema)`
It only takes 3 params. I am marking this constructor as @VivisbleForTesting

**Review by dsilvam — APPROVED** [review]

**Review by jdotcms — COMMENTED** [review]
cool

**Pull request merged**